Core/SqlStatementModel/ConstantCollectionExpression.cs

```diff
@@ -0,0 +1,108 @@
+﻿// This file is part of the re-linq project (relinq.codeplex.com)
+// Copyright (c) rubicon IT GmbH, www.rubicon.eu
+//
+// re-linq is free software; you can redistribute it and/or modify it under
+// the terms of the GNU Lesser General Public License as published by the
+// Free Software Foundation; either version 2.1 of the License,
+// or (at your option) any later version.
+//
+// re-linq is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with re-linq; if not, see http://www.gnu.org/licenses.
+//
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq.Expressions;
+using System.Text;
+using Remotion.Linq.SqlBackend.Utilities;
+using Remotion.Utilities;
+
+namespace Remotion.Linq.SqlBackend.SqlStatementModel
+{
+  /// <summary>
+  /// Expression representing an <see cref="ICollection"/>, an <see cref="ICollection{T}"/> or an <see cref="IReadOnlyCollection{T}"/>.
+  /// </summary>
+  public class ConstantCollectionExpression : Expression
+  {
+    public IEnumerable Collection { get; }
```

Comment:
I undestand why it's done this way, but I'm still allowed to feel like this is very very wrong, correct?

Reply:
It does feel weird. We could change the type to `ICollection`, but then we would have to wrap the original value in an `ArrayList` if its type does not implement `ICollection`. Should we do that?

Reply:
I considered this long and hard and the biggest reason agaisnt it is that transforming the actual value does not sound like the correct thing to do. I wonder if we should do the follwoing:
```
public object Collection {get;}
public IEnumerable Items {get;}
```

Reply:
I'm okay with `object Collection { get; }`, but I would suggest either:
* `IEnumerable GetItems();` while keeping `IsEmptyCollection { get; }`, or
* `ICollection GetItems();` with caching of the `ICollection` (which may be our wrapping `ArrayList`)